[FEATURE] API for efficient raster sampling

nyalldawson opened the pull request:
Add QgsRasterDataProvider::sample method for efficient sampling of rasters at a given point. This is an alternative to the ::identify method, which is less efficient but more powerful.

nyalldawson committed:
Less noisy identify method

nyalldawson committed:
Modernize code

nyalldawson committed:
[FEATURE] Add QgsRasterDataProvider::sample method for efficient
sampling of rasters at a given point

This is an alternative to the ::identify method, which is less
efficient but more powerful

nyalldawson committed:
Dox++

nyalldawson committed:
Unit tests for QgsRasterDataProvider::sample

nyalldawson committed:
More heavily optimised sample method for gdal provider

nyalldawson commented:
@rouault do you mind looking at the gdal provider implementation here? Are there any efficiency improvements you could suggest?

wonder-sk commented on src/core/raster/qgsrasterdataprovider.h:
```diff
@@ -355,10 +355,26 @@ class CORE_EXPORT QgsRasterDataProvider : public QgsDataProvider, public QgsRast
+     * is supported by provider.
+     *
+     * A null QVariant will be returned if the point is outside data source extent, or an invalid
+     * band number was specified.
```
Would be good to say what kind of types to expect in QVariant. Looks like currently we only support double... if that's always going to be the case, I would probably prefer to have double as return type + "bool *ok" to indicate success/failure

nyalldawson replied:
Fixed, I've removed the variant.

wonder-sk commented on src/providers/gdal/qgsgdalprovider.cpp:
```diff
@@ -1082,11 +1085,52 @@ QgsRasterIdentifyResult QgsGdalProvider::identify( const QgsPointXY &point, QgsR
 }
 
+bool QgsGdalProvider::worldToPixel( double x, double y, int &col, int &row ) const
+{
```
Can we have a small note for the reference that this solves
    Xgeo = GT(0) + Xpixel*GT(1) + Yline*GT(2)
    Ygeo = GT(3) + Xpixel*GT(4) + Yline*GT(5)
when Xgeo, Ygeo are given?

wonder-sk commented on src/providers/gdal/qgsgdalprovider.cpp:
```diff
@@ -1082,11 +1085,52 @@ QgsRasterIdentifyResult QgsGdalProvider::identify( const QgsPointXY &point, QgsR
+  if ( div < 2 * std::numeric_limits<double>::epsilon() )
+    return false;
+  double doubleCol = -( mGeoTransform[2] * ( mGeoTransform[3] - y ) + mGeoTransform[5] * x - mGeoTransform[0] * mGeoTransform[5] ) / div;
+  double doubleRow = ( mGeoTransform[1] * ( mGeoTransform[3] - y ) + mGeoTransform[4] * x - mGeoTransform[0] * mGeoTransform[4] ) / div;
```
A part of the equation could be simplified to ( x - mGeoTransform[0] ) * ( ... )

wonder-sk commented on src/providers/gdal/qgsgdalprovider.cpp:
```diff
@@ -1082,11 +1085,52 @@ QgsRasterIdentifyResult QgsGdalProvider::identify( const QgsPointXY &point, QgsR
+  double doubleCol = -( mGeoTransform[2] * ( mGeoTransform[3] - y ) + mGeoTransform[5] * x - mGeoTransform[0] * mGeoTransform[5] ) / div;
+  double doubleRow = ( mGeoTransform[1] * ( mGeoTransform[3] - y ) + mGeoTransform[4] * x - mGeoTransform[0] * mGeoTransform[4] ) / div;
+  col = static_cast< int >( doubleCol );
+  row = static_cast< int >( doubleRow );
```
Is it correct to round down here? (I don't know, just asking...)

nyalldawson replied:
Yes, it's correct - otherwise the values are 0.5 pixels off

wonder-sk commented on src/core/raster/qgsrasterdataprovider.h:
```diff
@@ -355,10 +355,26 @@ class CORE_EXPORT QgsRasterDataProvider : public QgsDataProvider, public QgsRast
+     * on the same zoom level as a displayed map and to do effective
+     * caching (WCS). If context params are not specified the highest
+     * resolution is used. capabilities() may be used to test if format
+     * is supported by provider.
```
The last sentence is only relevant to identify() I guess

wonder-sk reviewed (commented):
Cool stuff. I think ideally people should still be advised to use raster blocks to sample values with reasonable efficiency (if doing more than one sample) - but yeah, much better than doing identify...

wonder-sk commented:
> I agree, but I'm not sure how this approach could be used with something like #7393.

Yeah true...

nyalldawson committed:
Nicer API for raster sampling

nyalldawson committed:
Base class implementation of sample uses identify

This allows derived classes with optimised identify methods to
utilise the same optimisations for sample. Better would be for
them to implement optimised sample methods, but this is better
than using the unoptimised block approach for all providers.

nyalldawson committed:
Improvements and docs for worldToPixel

nyalldawson committed:
Use sample instead of identify in raster sampling alg

About 25x faster on large layers

nyalldawson committed:
Capitalization

nyalldawson committed:
Add missing override

nyalldawson commented on src/providers/gdal/qgsgdalprovider.cpp:
```diff
@@ -1082,11 +1085,65 @@ QgsRasterIdentifyResult QgsGdalProvider::identify( const QgsPointXY &point, QgsR
+};
+
+double QgsGdalProvider::sample( const QgsPointXY &point, int band, bool *ok, const QgsRectangle &, int, int, int )
+{
```
@rouault can you suggest any improvements here? Any way to eek out more performance from this method?
